src/app/AttributePathExpandIterator.h

```diff
@@ -20,78 +20,100 @@
+///      }
+///
+///   OR:
+///
+///      ConcreteAttributePath path;
+///      AttributePathExpandIterator iterator(mState);
+///
+///      while (iterator.Next(path)) {
+///         // use `path` here`
+///      }
+///
+/// USAGE requirements and assumptions:
+///
+///    - There should be only one single AttributePathExpandIterator for a state  at a time.
+///
+///    - `State` is automatically updated by the AttributePathExpandIterator, so
+///      calling `Next` on the iterator will update the state variable.
+///
+///
 class AttributePathExpandIterator
 {
 public:
-    AttributePathExpandIterator(DataModel::Provider * provider, SingleLinkedListNode<AttributePathParams> * attributePath);
-
-    /**
-     * Proceed the iterator to the next attribute path in the given cluster info.
-     *
-     * Returns false if AttributePathExpandIteratorDataModeDataModel has exhausted all paths in the given AttributePathParams list.
-     */
-    bool Next();
-
-    /**
-     * Fills the aPath with the path the iterator currently points to.
-     * Returns false if the iterator is not pointing to a valid path (i.e. it has exhausted the cluster info).
-     */
-    bool Get(ConcreteAttributePath & aPath)
+    class State
     {
-        aPath = mOutputPath;
-        return (mpAttributePath != nullptr);
-    }
+    public:
+        friend class AttributePathExpandIterator;
 
-    /**
-     * Reset the iterator to the beginning of current cluster if we are in the middle of expanding a wildcard attribute id for some
-     * cluster.
-     *
-     * When attributes are changed in the middle of expanding a wildcard attribute, we need to reset the iterator, to provide the
-     * client with a consistent state of the cluster.
-     */
-    void ResetCurrentCluster();
-
-    /** Start iterating over the given `paths` */
-    inline void ResetTo(SingleLinkedListNode<AttributePathParams> * paths)
-    {
-        *this = AttributePathExpandIterator(mDataModelProvider, paths);
-    }
+        /// External callers can only ever start iterating on a new path from the beginning
+        static State StartIterating(SingleLinkedListNode<AttributePathParams> * path) { return State(path); }
+
+        /// Copies are allowed
+        State(const State &)             = default;
+        State & operator=(const State &) = default;
+
+        State() : mAttributePath(nullptr) {}
+
+        /// Reset the iterator to the beginning of current cluster if we are in the middle of expanding a wildcard attribute id for
+        /// some cluster.
+        ///
+        /// When attributes are changed in the middle of expanding a wildcard attribute, we need to reset the iterator, to provide
+        /// the client with a consistent state of the cluster.
+        void IterateFromTheStartOfTheCurrentCluster()
+        {
+            VerifyOrReturn(mAttributePath != nullptr && mAttributePath->mValue.HasWildcardAttributeId());
+            mLastOutputPath.mAttributeId = kInvalidAttributeId;
```

Comment:
Suggest making new constants where `kNotYetStartedXxx == kInvalidXxx`. That keeps the marker value semantics without reading code that keeps looking for "invalid" values.

Reply:
I think `Invalid` is the correct name in this situation, like "we do not have a valid value at this point". I will document the invariants to make the code easier to understand if someone actually looks at it, however I would rather not over-alias the same constants with different names as that leads to confusion from other sides (without looking at definitions, if I see kA and kB, I generally assume `kA != kB` and if they are the same, rules of when to use kA vs kB when they end up with the same result are guaranteed to be eventually inconsistent.